### DISCUSSION TITLE
Can vocabsieve import KOreader highlights form a non kobo device?

### DISCUSSION THREAD

**Discussion by YashaNoos** [FreeLanguageTools/vocabsieve#49] [Q&A] [answered]
Hello,
I have an Onyx boox nova air with KOreader installed and I don't seem to be able to have vocabsieve import from or even simply access it. My device doesn't show up in the vocabsieve import window. 
![image](https://user-images.githubusercontent.com/109368517/215850550-69b4caf2-8e63-4a53-941b-3de891a23180.png)
I ask here because I don't know if it's a bug or vocabsieve isn't meant to work for non Kobo devices.

**Comment by Ceynou** [comment]
What do you mean it doesn't show up from vocabsieve, on your screenshot I don't see it in your windows manager either. All you should be needing is an epub its corresponding .sdr folder containing a Lua file, but I'm not sure if it's working right now, I tried with Koreader on my jailbroken kindle

Anyway if it's for japanese and you're on the same wifi as your PC, it's more interesting to read on ttu reader(on kiwi browser) with yomichan and [Ankiconnect android](https://github.com/KamWithK/AnkiconnectAndroid)

**Comment by YashaNoos** [comment]
My nova air does appear in the windows manager on the left but not in the vocabsieve import window.
<img width="372" alt="app_xFByT9tpqH" src="https://user-images.githubusercontent.com/109368517/215865083-6a8a2e4a-dfbc-4800-a999-1143f544ba63.png">
I tried to import on my pc my books folder containing my epubs and .sdr folder but I got this error when using the vocabsieve import window.
<img width="251" alt="app_lacCOxZ33T" src="https://user-images.githubusercontent.com/109368517/215865765-c8cdf891-0b9d-493e-804b-a43af901d8b7.png">
I have checked inside the lua of one of the books I have made highlights in and the highlight is present in it.

>     ["highlight"] = {
        [134] = {
            [1] = {
                ["pos1"] = "/body/DocFragment[14]/body/div/p[60]/text().48",
                ["datetime"] = "2023-01-31 20:34:47",
                ["chapter"] = "The Cat and the Skull",
                ["drawer"] = "lighten",
                ["text"] = "“That is naught but truth,” said Saremes calmly.",
                ["pos0"] = "/body/DocFragment[14]/body/div/p[60]/text().0",
            },
        },
    },

I already have a yomichan/ankiconnect setup on PC to study Japanese but for now I use my ereader to improve my English vocab.

**Reply by Ceynou** [reply]
Yeah I'm blind 🤦‍♂️
Same and I have the same error, the dev said he'll look into it, I gave him a sample, he's also looking for a way to implement importing lookups from the database(vocabulary builder like for the kindle)

**Reply by YashaNoos** [reply]
> Yeah I'm blind 🤦‍♂️

haha you had me double check!
And thanks for your answer, it's good to know they are working on it.

**Reply by 1over137** [reply]
Unlike the Kindle one the vocabulary builder in KOReader does not record context by default, so I'll probably have to still rely on going through the book to find the sentence.

**Comment by 1over137** [comment, marked as answer]
It should work just fine on a non-Kobo device. There is nothing specific to Kobo in the importer module.
If Windows doesn't show it on the select folder dialog, there isn't anything I can do about it. Since you have an Android device, you might consider using Syncthing to synchronize it with your computer. Try to also synchronize the KOReader settings folder, since it will be used later (though not obligatory) to log your lookups for tracking known vocabulary. 
This is probably a bug caused by me assuming the layout of an epub file incorrectly. I made a workaround for it. Try this build and see if it fixes the issue:
https://ci.appveyor.com/api/buildjobs/x45dwi5qb91h63w2/artifacts/cx_freeze%2Fwin%2Fdist%2FVocabSieve-v0.9.3-win64.msi

**Reply by YashaNoos** [reply]
Thanks for the quick response, Syncthing seems to be a good workaround I'll give it a try.
As for the new build, Vocabsieve don't give me an error window anymore but simply close itself down when I confirm the folder selection. I tried on other unrelated folders and the same thing happens.

**Reply by 1over137** [reply]
Seems like it's still broken then. Can you send the offending epub+.sdr? 
I'm considering using the vocabulary builder in KOReader for the context, but the fact that it's not on by default makes this troublesome. Though, that would be much more stable than parsing all the ebooks, which also has the issue of not supporting all formats.

**Reply by YashaNoos** [reply]
Here is one ebook with associated .sdr I hope it help.
But the problem seems to be that Vocabsieve crash when trying to select any folder.
As for using the vocabulary builder it would also be a good work flow improvement: looking up a word in the built in dictionary->Add to vocabulary builder, done. 
[Books.zip](https://github.com/FreeLanguageTools/vocabsieve/files/10553656/Books.zip)

**Reply by 1over137** [reply]
The crashing is actually an unrelated issue; it's caused by me removing the try;except blocks for debugging and forgetting to add it back.
With that said, there seems to be two actual problems. 1. You are selecting the whole sentence. You should only select the word needed. (meant to be done after you look it up, where a highlight button appears) 2. There is an unexpected issue with an index error. I'll try to fix this.

**Reply by 1over137** [reply]
Make sure to enable saving context in your Vocabulary Builder. The KOReader devs say it's been on by default for a while now, but those who had it installed before that would still have it off. I'm not sure I'll be able to solve your problem, but I'll create a new importing method for KOReader that would use the database directly, which should be much more reliable. For now, just use Vocabulary builder.

**Reply by YashaNoos** [reply]
Thanks for your time, I will give feed back once the version using Vocabulary Builder is available.

**Reply by 1over137** [reply]
It should now be available!
Notice that I switched to using Github Actions since Appveyor doesn't let me use multiple jobs on free tier.
You should select a folder that contains both KOReader settings and your books. For Kobo you can just select the root, and for Android probably the whole user folder. 
Builds here:
https://nightly.link/FreeLanguageTools/vocabsieve/workflows/build-binaries/master

**Reply by YashaNoos** [reply]
I just tried it and it works perfectly! 
I imported my whole settings folder from KOreader alongside my books and VocabSieve can work with that without problem. It's a huge improvement on my old workflow. Thanks for the hard work!

**Reply by 1over137** [reply]
> hard work

On the contrary, this is far easier to implement than the original KOReader import. I'm not so sure why I resorted to doing that without fully reading through the documentation about vocab builder. Should read more next time..
